pkg/querier/queryrange/detected_fields.go

```diff
@@ -0,0 +1,370 @@
+package queryrange
+
+import (
+	"context"
+	"net/http"
+	"slices"
+	"strconv"
+	"time"
+
+	"github.com/axiomhq/hyperloglog"
+	"github.com/dustin/go-humanize"
+	"github.com/grafana/dskit/httpgrpc"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/v3/pkg/logproto"
+	logql_log "github.com/grafana/loki/v3/pkg/logql/log"
+	"github.com/grafana/loki/v3/pkg/logql/syntax"
+	"github.com/grafana/loki/v3/pkg/logqlmodel"
+	"github.com/grafana/loki/v3/pkg/querier/plan"
+	base "github.com/grafana/loki/v3/pkg/querier/queryrange/queryrangebase"
+
+	"github.com/grafana/loki/pkg/push"
+)
+
+func NewDetectedFieldsHandler(
+	limitedHandler base.Handler,
+	logHandler base.Handler,
+	limits Limits,
+) base.Middleware {
+	return base.MiddlewareFunc(func(next base.Handler) base.Handler {
+		return base.HandlerFunc(
+			func(ctx context.Context, req base.Request) (base.Response, error) {
+				r, ok := req.(*DetectedFieldsRequest)
+				if !ok {
+					return next.Do(ctx, req)
+				}
+
+				resp, err := makeDownstreamRequest(ctx, limits, limitedHandler, logHandler, r)
+				if err != nil {
+					return nil, err
+				}
+
+				re, ok := resp.(*LokiResponse)
+				if !ok || re.Status != "success" {
+					return resp, nil
+				}
+
+				detectedFields := parseDetectedFields(r.FieldLimit, re.Data.Result)
+				fields := make([]*logproto.DetectedField, len(detectedFields))
+				fieldCount := 0
+				for k, v := range detectedFields {
+					p := v.parsers
+					if len(p) == 0 {
+						p = nil
+					}
+					fields[fieldCount] = &logproto.DetectedField{
+						Label:       k,
+						Type:        v.fieldType,
+						Cardinality: v.Estimate(),
+						Parsers:     p,
+					}
+
+					fieldCount++
+				}
+
+				return &DetectedFieldsResponse{
+					Response: &logproto.DetectedFieldsResponse{
+						Fields:     fields,
+						FieldLimit: r.GetFieldLimit(),
+					},
+					Headers: re.Headers,
+				}, nil
+			})
+	})
+}
+
+func makeDownstreamRequest(
+	ctx context.Context,
+	limits Limits,
+	limitedHandler, logHandler base.Handler,
+	req *DetectedFieldsRequest,
+) (base.Response, error) {
+	expr, err := syntax.ParseLogSelector(req.Query, true)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, "%s", err.Error())
+	}
+
+	if err := validateMaxEntriesLimits(ctx, req.LineLimit, limits); err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, "%s", err.Error())
+	}
+
+	if err := validateMatchers(ctx, limits, expr.Matchers()); err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, "%s", err.Error())
+	}
+
+	lokiReq := &LokiRequest{
+		Query:     req.GetQuery(),
+		Step:      req.GetStep(),
+		StartTs:   req.GetStartTs(),
+		EndTs:     req.GetEndTs(),
+		Direction: logproto.BACKWARD,
+		Limit:     req.GetLineLimit(),
+		Path:      "/loki/api/v1/query_range",
+	}
+
+	lokiReq.Plan = &plan.QueryPlan{
+		AST: expr,
+	}
+
+	if expr.HasFilter() {
+		return logHandler.Do(ctx, lokiReq)
+	}
+	return limitedHandler.Do(ctx, lokiReq)
```

Comment:
Responses seem to correctly contain structured metadata and parsed labels. Maybe injecting that header wasn't even necessary?

Reply:
Great point, I definitely overlooked that! I don't _think_ we need it, but I will double check. We are very dependent on the `Entry` having the correct `Parsed` fields populated, but I wonder if we're getting that from the parsers in the QF rather than in the Queriers? However, all the test cases assume they come parsed, so we might want that header for consistency.

Reply:
glad you brought this up @svennergr! Turns out the behavior of populating parsed fields is different in the ingester (ie. head block) compared to the store. Recent queries to the ingesters get entries that are always parsed, which is why our local results looked good. Once you get past that query ingesters within window, however, we need to set this to make sure it's entries are populated in the store.